pennylane/templates/layers/quantum_number_preserving_u2.py

```diff
@@ -0,0 +1,216 @@
+        :align: center
+        :width: 25%
+        :target: javascript:void(0);
+
+    |
+
+    The 2-parameter, 4-qubit :math:`\hat{Q}` gate is decomposed as follows:
+
+    |
+
+    .. figure:: ../../_static/templates/layers/quantum_number_preserving_decompos.png
+        :align: center
+        :width: 100%
+        :target: javascript:void(0);
+
+    |
+
+    Args:
+        weights (tensor_like): Array of weights of shape ``(L, D, 2)``\.
+            ``L`` is the number of gate fabric layers and :math:`D=M-1`
+            is the number of :math:`\hat{Q}(\varphi, \theta)` gates per layer.
+        wires (Iterable): wires that the template acts on
+        init_state (tensor_like): iterable or shape ``(len(wires),)`` tensor representing the Hartree-Fock state
+            used to initialize the wires
+        pi_gate_include (boolean): If `True`, sets the optional constant :math:`\Pi` to :math:`\text{QuantumNumberPreserving}_{OR}(\pi)`.
+            Default value is :math:`\hat{I}`.
+
+    .. UsageDetails::
+
+        #. The number of wires :math:`N` has to be equal to the number of
+           spin orbitals included in the active space.
+
+        #. The number of trainable parameters scales linearly with the number of layers as
+           :math:`2D(N-1)`.
+
+        An example of how to use this template is shown below:
+
+        .. code-block:: python
+
+            import pennylane as qml
+            from pennylane.templates import ParticleConservingU1
+            from functools import partial
+
+            # Build the electronic Hamiltonian from a local .xyz file
+            h, qubits = qml.qchem.molecular_hamiltonian("h2", "h2.xyz")
+
+            # Define the Hartree-Fock state
+            electrons = 2
+            ref_state = qml.qchem.hf_state(electrons, qubits)
+
+            # Define the device
+            dev = qml.device('default.qubit', wires=qubits)
+
+            # Define the ansatz
+            ansatz = partial(QuantumNumberPreservingU2, init_state=ref_state, pi_gate_include=True)
+
+            # Define the cost function
+            cost_fn = qml.ExpvalCost(ansatz, h, dev)
+
+            # Compute the expectation value of 'h'
+            layers = 2
+            params = qml.init.quantum_number_preserving_u2_normal(layers, qubits)
+            print(cost_fn(params))
+
+        **Parameter shape**
+
+        The shape of the weights argument can be computed by the static method
+        :meth:`~.QuantumNumberPreservingU2.shape` and used when creating randomly
+        initialised weight tensors:
+
+        .. code-block:: python
+
+            shape = QuantumNumberPreservingU2.shape(n_layers=2, n_wires=2)
+            weights = np.random.random(size=shape)
+
+
+    """
+    num_params = 1
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(
+        self, weights, wires=None, init_state=None, pi_gate_include=False, do_queue=True, id=None
```

Comment:
Why there are `None` defaults for  `wires` and `init_state`?

Reply:
That was a mistake. These are fixed now.